src/rulesets/Base/mapreduce.jl

```diff
@@ -19,6 +19,27 @@ function rrule(::typeof(sum), x::AbstractArray{T}; dims=:) where {T<:Number}
     return y, sum_pullback
 end
 
+function rrule(
+    config::RuleConfig{>:HasReverseMode}, ::typeof(sum), f, xs::AbstractArray; dims=:
+)
+    fx_and_pullbacks = map(x->rrule_via_ad(config, f, x), xs)
+    y = sum(first, fx_and_pullbacks; dims=dims)
+
+    pullbacks = last.(fx_and_pullbacks)
+    function sum_pullback(ȳ)
+        f̄_and_x̄s = [pullback(ȳ) for pullback in pullbacks]
```

Comment:
Have you investigated the performance of this? Storing an array of the need not be slow, but I wonder how well it works in practice?

I also wonder whether you could avoid making an array of tuples,  before `sum(first, fx_and_pullbacks)` and `last.(fx_and_pullbacks)` separates them. 

Finally, have you given thought to weird arrays, like `SMatrix`, or `CuArray`s? For the former at least, perhaps using `map` instead of a generator for `f̄_and_x̄s` may better preserve the structure.

Reply:
> Have you investigated the performance of this? Storing an array of the need not be slow, but I wonder how well it works in practice?

I have not. This is a first PR to get this feature out the door.
I am only really interested in it working, and having something to test AD's against.

For interest, Zygote's approach is
https://github.com/FluxML/Zygote.jl/blob/12f5c1d75eeaa8c7a818f2db7f8d082956c00cac/src/lib/array.jl#L296
which changes the primal to a `sum(f.x)` then AD's that expression.
Which hits https://github.com/FluxML/Zygote.jl/blob/12f5c1d75eeaa8c7a818f2db7f8d082956c00cac/src/lib/broadcast.jl#L172-L182
which is strictly worse than this, since it makes a temporary array for all the `y`s (which we just sum out), and for all the pullbacks

I will time them soon and post back, it will be interesting.

> I also wonder whether you could avoid making an array of tuples, before `sum(first, fx_and_pullbacks)` and `last.(fx_and_pullbacks)` separates them.

Yeah, I would like to look into that.
It's why i am doing `sum` rather than for `map` because I think for `map` that would be even more important.
It's a fiddly little pattern to write to make sure everything gets the right types, so I didn't want to do it for the first PR.

> Finally, have you given thought to weird arrays, like SMatrix, or CuArrays? For the former at least, perhaps using map instead of a generator for f̄_and_x̄s may better preserve the structure.

Good point, that's worth testing.

Reply:
Without AD, the reason to write `sum(f, x)` instead of `sum(f.(x))` is precisely to save allocations. It would be nice if the AD could preserve that, although possibly tricky.

I guess the size of the array of closures will depend on how much they capture, which may include both `x` and `y = f.(x)`. Ideally it would never include `x` since that already has its own array... and sometimes it would be quicker to re-calculate `y` but that seems even harder to arrange.

Reply:
I did a little experiment into using `mapreduce((net, pullbacks), (val, pb)) = (net + val, push!(pullbacks, pb))`
and I didn't get any performance improvement. Infact I got a large regression, but that could have been me messing stuff up.
I would like to do more testing of this.